pkg/reconciler/v1alpha1/autoscaling/kpa/kpa_scaler.go

```diff
@@ -107,57 +107,110 @@ func (ks *kpaScaler) Scale(ctx context.Context, pa *pav1alpha1.PodAutoscaler, de
+
+	if pa.Status.IsActivating() { // Active=Unknown
+		// Don't scale-to-zero during activation
+		desiredScale = scaleUnknown
+	} else if pa.Status.IsReady() { // Active=True
+		// Don't scale-to-zero if the PA is active
+
+		// Only let a revision be scaled to 0 if it's been active for at
+		// least the stable window's time.
+		if pa.Status.CanMarkInactive(config.StableWindow) {
+			return desiredScale, true
+		}
+		// Otherwise, scale down to 1 until the idle period elapses
+		desiredScale = 1
+	} else { // Active=False
+		// Don't scale-to-zero if the grace period hasn't elapsed
+		if !pa.Status.CanScaleToZero(config.ScaleToZeroGracePeriod) {
+			return desiredScale, true
+		}
+	}
+	return desiredScale, false
+}
+
+// Scale attempts to scale the given PA's target reference to the desired scale.
+func (ks *kpaScaler) Scale(ctx context.Context, pa *pav1alpha1.PodAutoscaler, desiredScale int32) (int32, error) {
+	logger := logging.FromContext(ctx)
+
+	if checkPAOwnership(ctx, pa) {
+		return desiredScale, nil
+	}
+
+	resource, err := getResource(ctx, pa)
+	if err != nil {
 		return desiredScale, err
 	}
-	resource := apis.KindToResource(gv.WithKind(pa.Spec.ScaleTargetRef.Kind)).GroupResource()
 	resourceName := pa.Spec.ScaleTargetRef.Name
 
-	// Identify the current scale.
 	scl, err := ks.scaleClientSet.Scales(pa.Namespace).Get(resource, resourceName)
 	if err != nil {
 		logger.Errorw(fmt.Sprintf("Resource %q not found", resourceName), zap.Error(err))
 		return desiredScale, err
 	}
+
 	currentScale := scl.Spec.Replicas
 
-	min, max := pa.ScaleBounds()
-	if newScale := applyBounds(min, max, desiredScale); newScale != desiredScale {
+	newScale := newScale(pa, desiredScale)
+	if newScale != desiredScale {
 		logger.Debugf("Adjusting desiredScale: %v -> %v", desiredScale, newScale)
 		desiredScale = newScale
 	}
 
-	if desiredScale == 0 {
-		// We should only scale to zero when both of the following conditions are true:
-		//   a) The PA has been active for atleast the stable window, after which it gets marked inactive
-		//   b) The PA has been inactive for atleast the grace period
-
-		config := ks.getAutoscalerConfig()
-
-		if pa.Status.IsActivating() { // Active=Unknown
-			// Don't scale-to-zero during activation
-			desiredScale = scaleUnknown
-		} else if pa.Status.IsReady() { // Active=True
-			// Don't scale-to-zero if the PA is active
-
-			// Only let a revision be scaled to 0 if it's been active for at
-			// least the stable window's time.
-			if pa.Status.CanMarkInactive(config.StableWindow) {
-				return desiredScale, nil
-			}
-			// Otherwise, scale down to 1 until the idle period elapses
-			desiredScale = 1
-		} else { // Active=False
-			// Don't scale-to-zero if the grace period hasn't elapsed
-			if !pa.Status.CanScaleToZero(config.ScaleToZeroGracePeriod) {
-				return desiredScale, nil
-			}
-		}
+	desiredScale, returnDesiredScale := ks.handleZeroScale(pa, desiredScale)
+	if returnDesiredScale {
```

Comment:
Hm, with the defaulting to 1 you don't need the duplication though, do you? You can just call `adjust...` and that'll return the correct value to scale to?

Reply:
Updated. Wdyt?

Reply:
I think my suggestions have just made it worse 😢 . Sorry about that. Shall we revert to your initial suggestion and just adjust the naming of the boolean variable to something like `applyDesiredScale` (which we'd need inverse from what it's now). Essentially this boolean is telling us whether or not we actually want to apply the scale returned by `handleZeroScale` (which I'd call `handleScaleToZero` since we're not at zero in that case).

Sorry for the confusion.

Reply:
Done.